### PR TITLE
Add typing to ImapHook

### PR DESCRIPTION
**Pull request by darwinyip**
---
This adds typing to ImapHook as part of https://github.com/apache/airflow/issues/9708

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by darwinyip** [comment]
For the following mypy errors:
```
providers/imap/hooks/imap.py:69: error: Incompatible types in assignment (expression has type "IMAP4_SSL", variable has type "None")
providers/imap/hooks/imap.py:70: error: "None" has no attribute "login"
providers/imap/hooks/imap.py:202: error: "None" has no attribute "select"
providers/imap/hooks/imap.py:214: error: "None" has no attribute "close"
providers/imap/hooks/imap.py:219: error: "None" has no attribute "search"
providers/imap/hooks/imap.py:224: error: "None" has no attribute "fetch"
```
It seems related to https://github.com/apache/airflow/blob/master/airflow/providers/imap/hooks/imap.py#L66-L69 which I found this issue https://github.com/python/mypy/issues/4520

I recommend adding `--strict-optional` flag.

File: airflow/providers/imap/hooks/imap.py [hunk 1]
```diff
@@ -65,12 +66,17 @@ def get_conn(self):
 
         if not self.mail_client:
             conn = self.get_connection(self.imap_conn_id)
-            self.mail_client = imaplib.IMAP4_SSL(conn.host)
-            self.mail_client.login(conn.login, conn.password)
+            self.mail_client = imaplib.IMAP4_SSL(conn.host)  # type:ignore
```
Comment (mik-laj) [inline airflow/providers/imap/hooks/imap.py]: Why don't we fix this problem?

File: airflow/providers/imap/hooks/imap.py [hunk 2]
```diff
@@ -41,7 +42,7 @@ class ImapHook(BaseHook):
     :type imap_conn_id: str
     """
 
-    def __init__(self, imap_conn_id='imap_default'):
+    def __init__(self, imap_conn_id: str = 'imap_default') -> None:
         super().__init__()
         self.imap_conn_id = imap_conn_id
         self.mail_client = None
```
Comment (mik-laj) [inline airflow/providers/imap/hooks/imap.py]:
```suggestion
        self.mail_client: Optional[imaplib.IMAP4_SSL] = None
```

Comment (darwinyip) [reply airflow/providers/imap/hooks/imap.py]: This still gives the same mypy errors.

Comment (mik-laj) [reply airflow/providers/imap/hooks/imap.py]: You need to add runtime-checks for type: See: https://github.com/apache/airflow/pull/9733
```python
if not self.cli:
    raise Exception("The 'cli' should be initialized before!")
```

**Review by mik-laj — APPROVED** [review]

**Pull request merged**